### PR TITLE
[Heartbeat] Log error on dupe monitor ID instead of strict req

### PR DESCRIPTION
**Pull request by andrewvc**
Currently Heartbeat will not start > 1 monitor with the same ID. This can create unintuitive scenarios in situations where a user
changes a monitor by adding a new entry via some reload mechanism (say k8s autodiscover) before the old one is deleted.

This PR changes the behavior to only log errors in this situation. Only one monitor will be active, but it will be the newest monitor.

There may be other fixes we should do in other areas, but this is probably a better failure method than what we currently have.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: heartbeat/monitors/dedup.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/dedup.go:44]: should we log like stopping previous monitor with same monitor id?

**Review by justinkambic — COMMENTED** [review]
One thing I'm noticing in my smoke testing of this patch is that, while it does let me run a config with multiple monitors sharing an ID, it doesn't seem to respect the usual `CTRL+C` I'd 
use to kill the process.

When attempting to run the same multi-shared-ID config on `master` I got the error message/no-op that is typical of HB up until now.

In the case of different monitors with the same ID it did seem to log them, per the included screenshot:

image

File: heartbeat/monitors/monitor_test.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/monitor_test.go:108]: can we also check if previous monitor is stopped?

**Review by vigneshshanmugam — APPROVED** [review]
Code changes LGTM. Added small comments.

File: heartbeat/monitors/dedup.go
Comment (jsoriano) [inline heartbeat/monitors/dedup.go:44]: Global logging functions should be avoided https://github.com/elastic/beats/issues/15699

File: heartbeat/monitors/monitor.go
Comment (jsoriano) [inline heartbeat/monitors/monitor.go:81]: Could this be part of the `RunnerFactory` to avoid needing a global object?

File: heartbeat/monitors/monitor_test.go
Comment (andrewvc) [reply heartbeat/monitors/monitor_test.go:108]: We do check for this at the end of the test by checking the number of stops invoked

File: heartbeat/monitors/monitor.go
Comment (jsoriano) [inline heartbeat/monitors/monitor.go:144]: `newMonitor` is used by `CheckConfig`. Checking configs shouldn't stop monitors existing monitors.

**Review by jsoriano — COMMENTED** [review]
Added some comments, mainly about avoiding the global objects if possible.

File: heartbeat/monitors/dedup.go
Comment (andrewvc) [reply heartbeat/monitors/dedup.go:44]: I don't know if that adds anything, because it's hard to tell them apart. Should I change the log message maybe?

File: heartbeat/monitors/dedup.go
Comment (andrewvc) [reply heartbeat/monitors/dedup.go]: Fixed!

File: heartbeat/monitors/monitor.go
Comment (andrewvc) [reply heartbeat/monitors/monitor.go]: Fixed this, it is now the case that `newMonitor` is side effect free

File: heartbeat/monitors/monitor.go
Comment (andrewvc) [reply heartbeat/monitors/monitor.go]: Great idea, done!

**Comment by andrewvc** [comment]
@justinkambic there was a deadlock issue that I've now fixed, that was blocking clean shutdowns, ctrl-c should now work, that's important!

File: heartbeat/beater/heartbeat.go
Comment (andrewvc) [inline heartbeat/beater/heartbeat.go:203]: Cleans everything up so that we only use a single factory instance in heartbeat

File: heartbeat/monitors/monitor.go
Comment (andrewvc) [inline heartbeat/monitors/monitor.go:134]: Prevents weird situations with dual stops, makes `stop()` idempotent.

**Comment by andrewvc** [comment]
OK, this should be ready for review again, it's largely rewritten, now using the factory for all state, and generally cleaning up how all the lifecycle stuff is handled. Thanks for the great input @jsoriano @justinkambic @vigneshshanmugam 

Races and deadlocks should all be gone.

File: heartbeat/monitors/factory.go
Comment (jsoriano) [inline heartbeat/monitors/factory.go:127]: Can this cause scheduling problems? For example if a monitor is scheduled every 30s, but for some reason an identical new one is created every 10s, replacing the old one, is it possible that it ends up being executed every 10s (or never?).

**Review by jsoriano — COMMENTED** [review]
LGTM if it looks good to uptime :slightly_smiling_face:

File: heartbeat/monitors/factory.go
Comment (andrewvc) [reply heartbeat/monitors/factory.go:127]: Great question. Fortunately that won't happen! The monitor object is bound to the scheduler [here](https://github.com/elastic/beats/blob/master/heartbeat/monitors/task.go#L87), so the way this works should be fine.

Comment (andrewvc) [reply heartbeat/monitors/factory.go:127]: Oh, I see your point, I guess my question is why would an identical new one be created every 10s? We do run monitors immediately on create. I think, in that situation, we need to explore why a new one, with a different config, is created every 10s. This patch isn't a substitute for config de-duplication I would say.

Comment (jsoriano) [reply heartbeat/monitors/factory.go:127]: > why would an identical new one be created every 10s?

It would be quite a corner case, in autodiscover for example if something that triggers autodiscover events is being periodically updated. Though in this case the issue would be there also without this change, as on update the config is already stopped an started.
So we are probably fine as this is now.

File: heartbeat/monitors/monitor.go
Comment (justinkambic) [inline heartbeat/monitors/monitor.go:221]:
```suggestion
// Stop the monitor without freeing it in global dedup
```
I know the function name is a proper noun here but it feels kinda redundant.

**Review by justinkambic — APPROVED** [review]
LGTM!

**Comment by awahab07** [comment]
**Post FF Testing**

LGTM
- Before the merge, heartbeat would error out adding a monitor with duplicate ID as an existing monitor.
- After the merge and rebuild, only warning was logged and the new monitor with duplicate ID overruled the previous one.

Before:
Screenshot 2021-11-23 at 15 11 51

After:
image
```
{"log.level":"warn","@timestamp":"2021-11-23T15:17:56.412+0100","log.logger":"monitor-factory","log.origin":{"file.name":"monitors/factory.go","file.line":125},"message":"monitor ID alibaba-monitor is configured for multiple monitors! IDs should be unique values, last seen config will win","service.name":"heartbeat","ecs.version":"1.6.0"}
```

|Previous monitor list|List after duplicate monitor is configured|
|:---:|:---:|
|Screenshot 2021-11-23 at 15 22 49|Screenshot 2021-11-23 at 15 21 43

**Pull request merged**
